src/components/MapController/MapController.js

```diff
@@ -89,93 +157,220 @@ var MapController = React.createClass({
   },
 
   componentWillReceiveProps: function (nextProps) {
-    this.selectedDataset = nextProps.meta[0];
-
-    this.requestTimeMetadata(this.selectedDataset.unique_id).then(response => {
+    var defaultDataset = nextProps.meta[0];
+    var run = defaultDataset.ensemble_member;
+    var start_date = defaultDataset.start_date;
+    var end_date = defaultDataset.end_date;
+    
+    var times = {};
+    var timesPromises = [];
+    
+    //set colours. In order of preference:
+    //1. colours received by prop
+    //2. colours set by the user or this function previously and held in state
+    //3. defaults (scalar rainbow if a single dataset, 
+    //             scalar greyscale and contours rainbow for 2)
+    var sPalette, cPalette;
+    if(nextProps.scalarPalette) {
+      sPalette = nextProps.scalarPalette;
+      cPalette = nextProps.contourPalette;
+    }
+    else if(this.state.scalarPalette) {
+      sPalette = this.state.scalarPalette;
+      cPalette = this.state.contourPalette;
+    }
+    else {
+      sPalette = nextProps.comparandMeta ? 'seq-Greys' : 'x-Occam';
+      cPalette = nextProps.comparandMeta ? 'x-Occam': undefined;
+    }
+   
+    //generate a list of timestamps available for the specified data
+    //by querying the metadata API.
+    var datasets = _.filter(nextProps.meta, 
+        {"ensemble_member": run, "start_date": start_date, "end_date": end_date });
+    
+    for(var i = 0; i < datasets.length; i++) {
+      timesPromises.push(this.requestTimeMetadata(datasets[i].unique_id));
+    }
+   
+    Promise.all(timesPromises).then(responses => {
+      for(var i = 0; i < responses.length; i++) {
+        var id = Object.keys(responses[i].data)[0];
+        for(var timeidx in responses[i].data[id].times) {
+          var idxString = JSON.stringify({
+            "timescale": responses[i].data[id].timescale,
+            "timeidx": timeidx,
+          });
+          times[idxString] = responses[i].data[id].times[timeidx];
+        }
+      }
+
+      //select a starting timestamp to display.
+      //should correspond to a 0th index, but might be annual, seasonal, or monthly
+      var startingTime = Object.keys(times)[0];
```

Comment:
... to here is AFAICT identical to most of `updateDataset` above. 

Factor that out into a shared function? If you don't, please apply the comment there about unordered hash keys here too.